src/common/mempool.cc

```diff
@@ -12,21 +12,156 @@
  *
  */
 
+#include <thread>
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-#if defined(_GNU_SOURCE) && defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
-#else
-// Thread local variables should save index, not &shard[index],
-// because shard[] is defined in the class
-static thread_local size_t thread_shard_index = mempool::num_shards;
-#endif
-
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
 // --------------------------------------------------------------
 
+namespace mempool {
+
+static size_t num_shard_bits;
+static size_t num_shards;
+
+static thread_local size_t thread_shard_index = max_shards;
+
+#if !defined(MEMPOOL_SCHED_GETCPU)
+// There is no cheap way of determining which core a thread is assigned to.
+// Statically assign threads to shards to minimize how many threads use the
+// same shard
+static int thread_shard_next;
+static std::mutex thread_shard_mtx;
+
+#else
+// Use sched_getcpu() to determine which core a thread is running on,
+// use this to assign the thread a shard and cache this value to
+// eliminate overhead of repeatedly calling sched_getcpu. Periodically
+// check the cached value to deal with non-bound threads switching cores.
+// Ideally this means that each CPU core accesses its own shard and there
+// are no shared CPU cachelines. If a thread switches cores performance
+// will be non-optimal (but atomics ensure correctness) until the switch
+// is noticed and a new shard is selected
+
+enum {
+  thread_shard_recheck_min_interval = 1,
+  thread_shard_recheck_max_interval = 10000
+};
+
+static thread_local int thread_shard_recheck_count;
+static thread_local int thread_shard_recheck_interval;
+
+// Keep statistics on how successfully we manage to track which core a thread
+// is running on
+//
+// correct_core - number of times thread to core binding is correctly tracked
+// incorrect_core - number of times thread to core binding is incorrectly tracked
+// checked_core - number of times sched_getcpu() has been called
+//
+// sched_getcpu_cost = 32 cycles
+// cacheline_ping_pong_cost = 150 cycles (intra socket), 400 cycles (inter socket)
+//
+// These costs will vary between CPU architectures, but relative difference between
+// costs is likely to be similar
+//
+// cached_algorithm_cost = checked_core * sched_getcpu_cost + incorrect_core * cacheline_ping_pong_cost
+// uncached_algorithm_cost = (correct_core + incorrect_core) * sched_getcpu_cost
+
+struct fault_stats_t {
+  ceph::atomic<size_t> correct_core = {0};
+  ceph::atomic<size_t> incorrect_core = {0};
+  ceph::atomic<size_t> checked_core = {0};
+
+  char __padding[128 - sizeof(ceph::atomic<size_t>)*2];
```

Comment:
shouldn't be `sizeof(ceph::atomic<size_t>)*3` since there are three atomics in the struct? Am I missing someting?

Reply:
Thats a bug - I'll fix it

Reply:
why is the structure being padded to 128 bytes?

Reply:
There is an array of fault_stats with one per shard with different shards being accessed by different threads. To avoid CPU cache lines being shared between cores the structure is made at least as big as a CPU cache line. See comments in mempool.h about choosing a cache line size - it is easier to use 128 (largest known CPU cache line size) than to try and work it out dynamically.

I noticed that C++17 defines std::hardware_destructive_interference_size which might be better than using 128, although I saw there were concerns with this because it is a compile time constant so causes problems if code is compiled for one specific processor type and then run on another with a different cache line size - see https://stackoverflow.com/questions/39680206/understanding-stdhardware-destructive-interference-size-and-stdhardware-cons

Reply:
I'm speculating the potentially-worrisome price for the ping-pong avoidance is decreased cache density. However, I lack number on that.

Reply:
The drop in cache density is usually worthwhile. Also remember that cache size (at least L1 and L2, but usually also L3) scales with number of cores.

If we start applying this technique in many places then it will be worthwhile grouping data per core together, but its not done enough to be a concern yet

Reply:
Commit https://github.com/ceph/ceph/pull/55696/commits/5f721120ab8433689b6c1601077c17f3755b471c resolves this issue by collecting stats for all the pools together. This means each shard consists of a small number of cache lines containing stats for all pools, and padding to cache line boundaries only happens once per shard, not once per shard per pool.